python/cudf/udf_cpp/shim.cu

```diff
@@ -662,3 +664,53 @@ make_definition_idx(BlockIdxMax, int64, int64_t);
 make_definition_idx(BlockIdxMax, float64, double);
 #undef make_definition_idx
 }
+
+/*
+NRT CUDA functions
+*/
+
+__device__ NRT_MemSys TheMSys = {
+  .allocator = {(NRT_malloc_func)malloc_wrapper, NULL, (NRT_free_func)free_wrapper},
+  .stats     = {0, 0, 0, 0}};
+
+__device__ void udf_str_dtor(void* udf_str, size_t size, void* dtor_info)
+{
+  auto ptr = reinterpret_cast<udf_string*>(udf_str);
+  ptr->~udf_string();
+}
+
+// Only used to allocate the right amount of space, see below
+struct meminfo_and_str {
+  NRT_MemInfo mi;
+  udf_string st;
+};
+
+/*
+Create a new MemInfo object holding the reference count of a udf_string. When returning
+new strings, shim functions expect a pointer to a stack allocated buffer into which it
+will construct the udf_string it returns. Since one can not safely build a MemInfo object
+around this stack memory, we store a copy of the udf_string itself next to the MemInfo
+which manages the lifetime of the udf_string.
+*/
+extern "C" __device__ int meminfo_from_new_udf_str(void** nb_retval, void* udf_str)
+{
+  // allocate enough room for both the meminfo and udf_string
+  meminfo_and_str* mi_and_str = (meminfo_and_str*)NRT_Allocate(sizeof(meminfo_and_str));
+  if (mi_and_str != NULL) {
+    auto mi_ptr        = &(mi_and_str->mi);
+    udf_string* st_ptr = &(mi_and_str->st);
+
+    // We pass a null size here because the udf_string actually exists on the stack
+    // and tracks the size of the string data that it points to.
+    NRT_MemInfo_init(mi_ptr, st_ptr, NULL, udf_str_dtor, NULL);
+
+    // copy the udf_string to the extra heap space
+    udf_string* in_str_ptr = reinterpret_cast<udf_string*>(udf_str);
+    memcpy(st_ptr, in_str_ptr, sizeof(udf_string));
```

Comment:
Following through the code in later PRs, it looks like the placement new is no longer needed (in functions like `replace()` etc in `udf_cpp/shim.cu` because the `udf_string` pointed-to by `st_ptr` was constructed by the 

```cuda
__device__ inline udf_string replace(string_view source,
                                     string_view target,
                                     string_view replacement)
```

function in `replace.cuh`, and overwrites the data that would have been put there by the placement new - @brandon-b-miller Do you see this flow through the code in the code in its entirety in #13181?

Reply:
> 



> IIUC `udf_str` will have been constructed with a placement `new` where the call to it originates from Numba-generated code when lowering functions like `replace()`. Would it help in the context of this PR if the comment / documentation were explicit that `udf_str` should be pointing to a fully-initialized `udf_string`?

I think I'm happy that `udf_str` is fully initialised. But is `*st_ptr` a valid `udf_string` object after we've just done a `memcpy` into it? If it were a simple C struct I think the answer would be yes, but it's not a simple C struct I think.

Reply:
Yes, it is valid.

Reply:
Thanks @davidwendt.